### PR TITLE
feat: add subscription managment

### PR DESCRIPTION
**Pull request by ISulimanov**
Fix: https://github.com/casdoor/casdoor/issues/1711

Demo site:
http://95.161.157.78/ (admin/123)

Pricing page (should be opened in incognito)
http://95.161.157.78/select-plan/pricing_2023_eu

image

Realizing that the PR is quite large, I tried to include only the minimum required functionality in it (MVP)

That the reason why only one flow is supported so far

**Supported:**

- Select plan => Sign up. Trial period is required unlike credit card.

**In future**

- Select plan => Pay => Sign up. In case of when No Trial and credit card will be required.

In few weeks I will provide docs.
In case of any bug in this feature I will do my best to fix it.

Thanks.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@ISulimanov plz fix CI:

image

**Comment by ISulimanov** [comment]
> @ISulimanov plz fix CI:
> 
> image

hope it should be fixed, I've used gofumpt for checking changes

File: object/migrator_1_317_0_PR_1857.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import (
+	"errors"
+
+	"github.com/xorm-io/xorm"
+	"github.com/xorm-io/xorm/migrate"
+)
+
+type Migrator_1_317_0_PR_1857 struct{}
+
+func (*Migrator_1_317_0_PR_1857) IsMigrationNeeded() bool {
```
Comment (hsluoyz) [inline object/migrator_1_317_0_PR_1857.go:26]: Table is auto created by adapter.go, no need to add a migrator?

```go
err = a.Engine.Sync2(new(User))
	if err != nil {
		panic(err)
	}

	err = a.Engine.Sync2(new(Role))
	if err != nil {
		panic(err)
	}

	err = a.Engine.Sync2(new(Permission))
	if err != nil {
		panic(err)
	}

	err = a.Engine.Sync2(new(Model))
	if err != nil {
		panic(err)
	}
```

File: web/src/backend/PricingBackend.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+// Copyright 2022 The Casdoor Authors. All Rights Reserved.
```
Comment (hsluoyz) [inline web/src/backend/PricingBackend.js:1]: 2023, check all

File: object/plan.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+package object
+
+import (
+	"fmt"
+
+	"github.com/casdoor/casdoor/util"
+	"github.com/xorm-io/core"
+)
+
+type PlanWithOptions struct {
```
Comment (hsluoyz) [inline object/plan.go:24]: No need to define `PlanWithOptions`, just add it to Plan:

Options []string `xorm:"-" json:"options"`

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/account.go [hunk 4]
```diff
@@ -189,6 +189,29 @@ func (c *ApiController) Signup() {
 	object.DisableVerificationCode(authForm.Email)
 	object.DisableVerificationCode(checkPhone)
 
+	signupFromPricing := len(authForm.Plan) > 0 && len(authForm.Pricing) > 0
```
[the remark this answers is not in the document]
Comment (ISulimanov) [reply controllers/account.go:192]: please take a look new Subscribe dedicated function

File: object/migrator_1_317_0_PR_1857.go (hunk 1 above)
Comment (ISulimanov) [reply object/migrator_1_317_0_PR_1857.go]: Thanks, looks like migration really not needed I've add new tables to createTable func

File: web/src/backend/PricingBackend.js (hunk 2 above)
Comment (ISulimanov) [reply web/src/backend/PricingBackend.js]: done

File: object/plan.go (hunk 3 above)
Comment (ISulimanov) [reply object/plan.go]: Thx, done

File: object/migrator.go [hunk 5]
```diff
@@ -27,6 +27,7 @@ func DoMigration() {
 		&Migrator_1_235_0_PR_1530{},
 		&Migrator_1_240_0_PR_1539{},
 		&Migrator_1_314_0_PR_1841{},
+		&Migrator_1_317_0_PR_1857{},
```
[the remark this answers is not in the document]
Comment (ISulimanov) [reply object/migrator.go]: migration removed

File: controllers/plan.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+func (c *ApiController) GetPlans() {
+	owner := c.Input().Get("owner")
+	limit := c.Input().Get("pageSize")
+	page := c.Input().Get("p")
+	field := c.Input().Get("field")
+	value := c.Input().Get("value")
+	sortField := c.Input().Get("sortField")
+	sortOrder := c.Input().Get("sortOrder")
+	if limit == "" || page == "" {
+
```
[the remark this answers is not in the document]
Comment (ISulimanov) [reply controllers/plan.go]: done

File: controllers/plan.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
+// Copyright 2022 The Casdoor Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (ISulimanov) [reply controllers/plan.go]: done

**Review by hsluoyz — APPROVED** [review]

**Comment by hsluoyz** [comment]
@ISulimanov hi, the PR has been merged. Can you add the docs?

**Comment by ISulimanov** [comment]
> @ISulimanov hi, the PR has been merged. Can you add the docs?

Hi, thx for good news, I will provide docs soon.

**Comment by amaurybrisou** [comment]
Hi, is the Paying step coming soon in your roadmap ? Thanks for your work !

**Comment by ISulimanov** [comment]
> Hi, is the Paying step coming soon in your roadmap ? Thanks for your work !

Hi many thanks for your feedback.
Hopefully soon I will able to provide next payment step with payment providers integration.

**Pull request merged**